File: services/server/src/safe_socket/safe_socket.py
import socket
from weakref import WeakKeyDictionary

MESSAGE_TYPE_SIZE = 1
PAYLOAD_LENGTH_SIZE = 4
MESSAGE_HEADER_SIZE = MESSAGE_TYPE_SIZE + PAYLOAD_LENGTH_SIZE

MAX_MESSAGE_TYPE = 255
MAX_PAYLOAD_SIZE = 16 * 1024 * 1024
MAX_EMPTY_OPERATIONS = 100

_RECEIVE_BUFFERS: WeakKeyDictionary[socket.socket, bytearray] = WeakKeyDictionary()


def recv_all(sock: socket.socket, size: int) -> bytes:
    if size < 0:
        raise ValueError(f"receive size must be non-negative: {size}")
    if size == 0:
        return b""

    received = _RECEIVE_BUFFERS.pop(sock, bytearray())

    while len(received) < size:
        chunk = sock.recv(size)
        if not chunk:
            raise ConnectionError("connection closed while receiving data")
        received.extend(chunk)

    result = bytes(received[:size])
    del received[:size]
    if received:
        _RECEIVE_BUFFERS[sock] = received

    return result


def send_all(sock: socket.socket, data: bytes) -> None:
    bytes_sent = 0
    empty_writes = 0

    while bytes_sent < len(data):
        sent = sock.send(data[bytes_sent:])
        if sent == 0:
            empty_writes += 1
            if empty_writes >= MAX_EMPTY_OPERATIONS:
                raise ConnectionError("socket made no progress while sending data")
            continue

        empty_writes = 0
        bytes_sent += sent


def send_message(sock: socket.socket, message_type: int, payload: bytes) -> None:
    if not 0 <= message_type <= MAX_MESSAGE_TYPE:
        raise ValueError(f"invalid message type: {message_type}")
    if len(payload) > MAX_PAYLOAD_SIZE:
        raise ValueError(f"payload exceeds {MAX_PAYLOAD_SIZE} bytes")

    header = bytes([message_type]) + len(payload).to_bytes(
        PAYLOAD_LENGTH_SIZE, byteorder="big"
    )
    send_all(sock, header)
    send_all(sock, payload)


def recv_message(sock: socket.socket) -> tuple[int, bytes]:
    header = recv_all(sock, MESSAGE_HEADER_SIZE)
    message_type = header[0]
    payload_size = int.from_bytes(header[MESSAGE_TYPE_SIZE:], byteorder="big")
    if payload_size > MAX_PAYLOAD_SIZE:
        raise ValueError(f"payload exceeds {MAX_PAYLOAD_SIZE} bytes")
    payload = recv_all(sock, payload_size)
    return message_type, payload

File: services/client/src/safe_socket/safe_socket.go
package safe_socket

import (
	"encoding/binary"
	"fmt"
	"io"
)

const (
	messageTypeSize    = 1
	payloadLengthSize  = 4
	messageHeaderSize  = messageTypeSize + payloadLengthSize
	maxPayloadSize     = 16 * 1024 * 1024
	maxEmptyOperations = 100
)

const MessageHeaderSize = messageHeaderSize

type MessageType byte

const (
	MessageTypeBet       MessageType = 1
	MessageTypeEnd       MessageType = 2
	MessageTypeWinner    MessageType = 3
	MessageTypeError     MessageType = 4
	MessageTypeBetsBatch MessageType = 5
	MessageTypeBatchAck  MessageType = 6
)

type Message struct {
	Type    MessageType
	Payload []byte
}

func SendAll(socket io.Writer, bytes []byte) error {
	bytesSent := 0
	emptyWrites := 0

	for bytesSent < len(bytes) {
		n, err := socket.Write(bytes[bytesSent:])
		bytesSent += n

		if err != nil {
			return err
		}
		if n == 0 {
			emptyWrites++
			if emptyWrites >= maxEmptyOperations {
				return io.ErrNoProgress
			}
			continue
		}
		emptyWrites = 0
	}

	return nil
}

func RecvAll(socket io.Reader, size int) ([]byte, error) {
	if size < 0 {
		return nil, fmt.Errorf("receive size must be non-negative: %d", size)
	}

	buffer := make([]byte, size)
	if err := recvAllInto(socket, buffer); err != nil {
		return nil, err
	}
	return buffer, nil
}

func recvAllInto(socket io.Reader, buffer []byte) error {
	bytesRead := 0
	emptyReads := 0

	for bytesRead < len(buffer) {
		n, err := socket.Read(buffer[bytesRead:])
		bytesRead += n

		if bytesRead == len(buffer) {
			return nil
		}
		if err != nil {
			return err
		}
		if n == 0 {
			emptyReads++
			if emptyReads >= maxEmptyOperations {
				return io.ErrNoProgress
			}
			continue
		}
		emptyReads = 0
	}

	return nil
}

func SendMessageWithHeader(socket io.Writer, message Message, header []byte) error {
	if !message.Type.isValid() {
		return fmt.Errorf("invalid message type %d", message.Type)
	}
	if len(message.Payload) > maxPayloadSize {
		return fmt.Errorf("payload exceeds %d bytes", maxPayloadSize)
	}
	if len(header) < messageHeaderSize {
		return fmt.Errorf("message header buffer must contain at least %d bytes", messageHeaderSize)
	}

	header[0] = byte(message.Type)
	binary.BigEndian.PutUint32(header[1:], uint32(len(message.Payload)))

	if err := SendAll(socket, header[:messageHeaderSize]); err != nil {
		return err
	}

	return SendAll(socket, message.Payload)
}

func RecvMessageInto(socket io.Reader, header []byte, payload []byte) (Message, error) {
	if len(header) < messageHeaderSize {
		return Message{}, fmt.Errorf("message header buffer must contain at least %d bytes", messageHeaderSize)
	}
	if err := recvAllInto(socket, header[:messageHeaderSize]); err != nil {
		return Message{}, err
	}

	messageType := MessageType(header[0])
	if !messageType.isValid() {
		return Message{}, fmt.Errorf("invalid message type %d", messageType)
	}
	payloadSize := int(binary.BigEndian.Uint32(header[1:]))
	if payloadSize > maxPayloadSize {
		return Message{}, fmt.Errorf("payload exceeds %d bytes", maxPayloadSize)
	}
	if cap(payload) < payloadSize {
		payload = make([]byte, payloadSize)
	} else {
		payload = payload[:payloadSize]
	}
	if err := recvAllInto(socket, payload); err != nil {
		return Message{}, err
	}

	return Message{
		Type:    messageType,
		Payload: payload,
	}, nil
}

func (messageType MessageType) isValid() bool {
	return messageType >= MessageTypeBet && messageType <= MessageTypeBatchAck
}
